✘ [ERROR] The symbol "__dirname" has already been declared

### Describe the bug

my vite.config.js
```js
import path from "path";
import { defineConfig } from "vite";
import { svelte } from "@sveltejs/vite-plugin-svelte";
import preprocess from "svelte-preprocess";

const __filename = import.meta.url;
const __dirname = path.dirname(__filename);

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [
    svelte({
      preprocess: preprocess(),
    }),
  ],
  resolve: {
    alias: {
      "@views": path.resolve(__dirname, "src/views"),
      "@components": path.resolve(__dirname, "src/components"),
      "@/assets": path.resolve(__dirname, "src/assets"),
      "@utils": path.resolve(__dirname, "src/utils"),
      "@hooks": path.resolve(__dirname, "src/hooks"),
      "@context": path.resolve(__dirname, "src/context"),
    },
  },
});
```

if I run npm run dev I get "✘ [ERROR] The symbol "__dirname" has already been declared" error
if I remove type: "module" from package.json everything works fine

### Reproduction

npm run dev

### System Info

```shell
System:
    OS: Linux 5.18 Arch Linux
    CPU: (4) x64 Intel(R) Core(TM) i5 CPU       M 560  @ 2.67GHz
    Memory: 1.18 GB / 3.76 GB
    Container: Yes
    Shell: 5.1.16 - /bin/bash
  Binaries:
    Node: 18.5.0 - /usr/bin/node
    npm: 8.13.2 - /usr/bin/npm
  Browsers:
    Brave Browser: 103.1.41.96
    Chromium: 103.0.5060.114
    Firefox: 102.0.1
```


### Used Package Manager

npm

### Logs

```shell
✘ [ERROR] The symbol "__filename" has already been declared

    vite.config.js:6:6:
      6 │ const __filename = import.meta.url;
        ╵       ~~~~~~~~~~

  The symbol "__filename" was originally declared here:

    vite.config.js:1:62:
      1 │ ...edov/Web/Svelte/Learn";const __filename = "/home/mirzaahmedov/Web/S...
        ╵                                 ~~~~~~~~~~

✘ [ERROR] The symbol "__dirname" has already been declared

    vite.config.js:7:6:
      7 │ const __dirname = path.dirname(__filename);
        ╵       ~~~~~~~~~

  The symbol "__dirname" was originally declared here:

    vite.config.js:1:6:
      1 │ const __dirname = "/home/mirzaahmedov/Web/Svelte/Learn";const __filena...
        ╵       ~~~~~~~~~

failed to load config from /home/mirzaahmedov/Web/Svelte/Learn/vite.config.js
error when starting dev server:
Error: Build failed with 2 errors:
vite.config.js:6:6: ERROR: The symbol "__filename" has already been declared
vite.config.js:7:6: ERROR: The symbol "__dirname" has already been declared
    at failureErrorWithLog (/home/mirzaahmedov/Web/Svelte/Learn/node_modules/esbuild/lib/main.js:1605:15)
    at /home/mirzaahmedov/Web/Svelte/Learn/node_modules/esbuild/lib/main.js:1251:28
    at runOnEndCallbacks (/home/mirzaahmedov/Web/Svelte/Learn/node_modules/esbuild/lib/main.js:1034:63)
    at buildResponseToResult (/home/mirzaahmedov/Web/Svelte/Learn/node_modules/esbuild/lib/main.js:1249:7)
    at /home/mirzaahmedov/Web/Svelte/Learn/node_modules/esbuild/lib/main.js:1358:14
    at /home/mirzaahmedov/Web/Svelte/Learn/node_modules/esbuild/lib/main.js:666:9
    at handleIncomingPacket (/home/mirzaahmedov/Web/Svelte/Learn/node_modules/esbuild/lib/main.js:763:9)
    at Socket.readFromStdout (/home/mirzaahmedov/Web/Svelte/Learn/node_modules/esbuild/lib/main.js:632:7)
    at Socket.emit (node:events:537:28)
    at addChunk (node:internal/streams/readable:324:12)
```

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitejs.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitejs/vite/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to [vuejs/core](https://github.com/vuejs/core) instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/packages/vite/src/node/config.ts
+++ b/packages/vite/src/node/config.ts
@@ -27,7 +27,6 @@ import {
   dynamicImport,
   isExternalUrl,
   isObject,
-  isTS,
   lookupFile,
   mergeAlias,
   mergeConfig,
@@ -816,7 +815,6 @@ export async function loadConfigFromFile(
   const getTime = () => `${(performance.now() - start).toFixed(2)}ms`
 
   let resolvedPath: string | undefined
-  let dependencies: string[] = []
 
   if (configFile) {
     // explicit config path is always resolved from cwd
@@ -852,42 +850,13 @@ export async function loadConfigFromFile(
   }
 
   try {
-    let userConfig: UserConfigExport | undefined
-
-    if (isESM) {
-      const fileUrl = pathToFileURL(resolvedPath)
-      const bundled = await bundleConfigFile(resolvedPath, true)
-      dependencies = bundled.dependencies
-
-      if (isTS(resolvedPath)) {
-        // before we can register loaders without requiring users to run node
-        // with --experimental-loader themselves, we have to do a hack here:
-        // bundle the config file w/ ts transforms first, write it to disk,
-        // load it with native Node ESM, then delete the file.
-        fs.writeFileSync(resolvedPath + '.mjs', bundled.code)
-        try {
-          userConfig = (await dynamicImport(`${fileUrl}.mjs?t=${Date.now()}`))
-            .default
-        } finally {
-          fs.unlinkSync(resolvedPath + '.mjs')
-        }
-        debug(`TS + native esm config loaded in ${getTime()}`, fileUrl)
-      } else {
-        // using Function to avoid this from being compiled away by TS/Rollup
-        // append a query so that we force reload fresh config in case of
-        // server restart
-        userConfig = (await dynamicImport(`${fileUrl}?t=${Date.now()}`)).default
-        debug(`native esm config loaded in ${getTime()}`, fileUrl)
-      }
-    }
-
-    if (!userConfig) {
-      // Bundle config file and transpile it to cjs using esbuild.
-      const bundled = await bundleConfigFile(resolvedPath)
-      dependencies = bundled.dependencies
-      userConfig = await loadConfigFromBundledFile(resolvedPath, bundled.code)
-      debug(`bundled config file loaded in ${getTime()}`)
-    }
+    const bundled = await bundleConfigFile(resolvedPath, isESM)
+    const userConfig = await loadConfigFromBundledFile(
+      resolvedPath,
+      bundled.code,
+      isESM
+    )
+    debug(`bundled config file loaded in ${getTime()}`)
 
     const config = await (typeof userConfig === 'function'
       ? userConfig(configEnv)
@@ -898,7 +867,7 @@ export async function loadConfigFromFile(
     return {
       path: normalizePath(resolvedPath),
       config,
-      dependencies
+      dependencies: bundled.dependencies
     }
   } catch (e) {
     createLogger(logLevel).error(
@@ -911,7 +880,7 @@ export async function loadConfigFromFile(
 
 async function bundleConfigFile(
   fileName: string,
-  isESM = false
+  isESM: boolean
 ): Promise<{ code: string; dependencies: string[] }> {
   const importMetaUrlVarName = '__vite_injected_original_import_meta_url'
   const result = await build({
@@ -954,7 +923,7 @@ async function bundleConfigFile(
               )};`
 
             return {
-              loader: isTS(args.path) ? 'ts' : 'js',
+              loader: args.path.endsWith('ts') ? 'ts' : 'js',
               contents: injectValues + contents
             }
           })
@@ -976,22 +945,38 @@ interface NodeModuleWithCompile extends NodeModule {
 const _require = createRequire(import.meta.url)
 async function loadConfigFromBundledFile(
   fileName: string,
-  bundledCode: string
-): Promise<UserConfig> {
-  const realFileName = fs.realpathSync(fileName)
-  const defaultLoader = _require.extensions['.js']
-  _require.extensions['.js'] = (module: NodeModule, filename: string) => {
-    if (filename === realFileName) {
-      ;(module as NodeModuleWithCompile)._compile(bundledCode, filename)
-    } else {
-      defaultLoader(module, filename)
+  bundledCode: string,
+  isESM: boolean
+): Promise<UserConfigExport> {
+  // for esm, before we can register loaders without requiring users to run node
+  // with --experimental-loader themselves, we have to do a hack here:
+  // write it to disk, load it with native Node ESM, then delete the file.
+  if (isESM) {
+    const fileUrl = pathToFileURL(fileName)
+    fs.writeFileSync(fileName + '.mjs', bundledCode)
+    try {
+      return (await dynamicImport(`${fileUrl}.mjs?t=${Date.now()}`)).default
+    } finally {
+      fs.unlinkSync(fileName + '.mjs')
     }
   }
-  // clear cache in case of server restart
-  delete _require.cache[_require.resolve(fileName)]
-  const raw = _require(fileName)
-  _require.extensions['.js'] = defaultLoader
-  return raw.__esModule ? raw.default : raw
+  // for cjs, we can register a custom loader via `_require.extensions`
+  else {
+    const realFileName = fs.realpathSync(fileName)
+    const defaultLoader = _require.extensions['.js']
+    _require.extensions['.js'] = (module: NodeModule, filename: string) => {
+      if (filename === realFileName) {
+        ;(module as NodeModuleWithCompile)._compile(bundledCode, filename)
+      } else {
+        defaultLoader(module, filename)
+      }
+    }
+    // clear cache in case of server restart
+    delete _require.cache[_require.resolve(fileName)]
+    const raw = _require(fileName)
+    _require.extensions['.js'] = defaultLoader
+    return raw.__esModule ? raw.default : raw
+  }
 }
 
 export function getDepOptimizationConfig(

--- a/packages/vite/src/node/utils.ts
+++ b/packages/vite/src/node/utils.ts
@@ -1130,8 +1130,6 @@ export function stripBomTag(content: string): string {
   return content
 }
 
-export const isTS = (filename: string): boolean => /\.[cm]?ts$/.test(filename)
-
 const windowsDrivePathPrefixRE = /^[A-Za-z]:[/\\]/
 
 /**

--- a/playground/ssr-vue/vite.config.js
+++ b/playground/ssr-vue/vite.config.js
@@ -10,6 +10,11 @@ const nestedVirtualId = '\0' + nestedVirtualFile
 
 const base = '/test/'
 
+// preserve this to test loading __filename & __dirname in ESM as Vite polyfills them.
+// if Vite incorrectly load this file, node.js would error out.
+globalThis.__vite_test_filename = __filename
+globalThis.__vite_test_dirname = __dirname
+
 export default defineConfig(({ command, ssrBuild }) => ({
   base,
   plugins: [

